File: arithmatic slices.java
//Timecomplexity:- O(n);
//Spaecomplexity:- O(1).




class Solution {
    public int numberOfArithmeticSlices(int[] nums) {
        int i=nums.length-3;
        int prev=0;
        int sum=0;
        while(i>=0){
            if(isArithmetic(nums[i],nums[i+1],nums[i+2])){
               prev=1+prev;
               sum=sum+prev;
            }
            else{
                prev=0;
            }
            i--;            //Starting from nums.length-3,if last three are valid adding one to prev value.
            // and sum would be adding previous subsets+present subsets which is total sum.
            //if not valid prev will be no altered.
            
        }
        return sum;
    }
    private boolean isArithmetic(int x,int y, int z){
        if((y-x)==(z-y)){
            return true;
        }
        return false;
    }
}

File: ArithmaticSlices.py
'''
Time Complexity --> O(n)
Space Complexity --> O(1)
'''
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        # #brute Force
        # count = 0
        # for i in range(len(nums)-1):
        #     for j in range(i+1, len(nums)-1):
        #         #print(nums[j]-nums[j-1])
        #         if nums[j]-nums[j-1] == nums[j+1]-nums[j]:
        #             count+=1
        #         else:
        #             break
                
        # return count
        prev = 0
        tot = 0
        for i in range(len(nums)-3, -1, -1):
            if nums[i+2]-nums[i+1] == nums[i+1]-nums[i]:
                prev+=1
                tot+=prev
            else:
                prev = 0
        return(tot)

                
